compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,329 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.h"
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h"
+#include "iree/compiler/Codegen/Common/VectorLayoutAnalysis.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/Verifier.h"
+#include "mlir/Rewrite/PatternApplicator.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-vector-distribution"
+
+using namespace mlir::iree_compiler::IREE::VectorExt;
+
+namespace mlir::iree_compiler {
+
+using VectorValue = TypedValue<VectorType>;
+
+struct DistributionSignature {
+  SmallVector<VectorLayoutInterface> operands;
+  SmallVector<VectorLayoutInterface> results;
+};
+
+static const char *kVectorLayoutFetcherStorageAttrName =
+    "__vector_layout_fetcher_storage";
+
+static void setOpSignature(Operation *op, VectorLayoutAnalysis &analysis) {
+  SmallVector<Attribute> operands;
+  SmallVector<Attribute> results;
+
+  for (Value operand : op->getOperands()) {
+    if (auto vectorOperand = dyn_cast<VectorValue>(operand)) {
+      operands.push_back(
+          analysis.getLayout<VectorLayoutInterface>(vectorOperand));
+      continue;
+    }
+    operands.push_back(VectorLayoutInterface());
+  }
+
+  for (Value result : op->getResults()) {
+    if (auto vectorResult = dyn_cast<VectorValue>(result)) {
+      results.push_back(
+          analysis.getLayout<VectorLayoutInterface>(vectorResult));
+      continue;
+    }
+    results.push_back(VectorLayoutInterface());
+  }
+
+  ArrayAttr operandsAttr = ArrayAttr::get(op->getContext(), operands);
+  ArrayAttr resultsAttr = ArrayAttr::get(op->getContext(), results);
+  SmallVector<Attribute> signature = {operandsAttr, resultsAttr};
+  op->setAttr(kVectorLayoutFetcherStorageAttrName,
+              ArrayAttr::get(op->getContext(), signature));
+}
+
+static bool hasOpSignature(Operation *op) {
+  return op->hasAttrOfType<ArrayAttr>(kVectorLayoutFetcherStorageAttrName);
+}
+
+static DistributionSignature getOpSignature(Operation *op) {
+  ArrayAttr signatureAttr =
+      op->getAttrOfType<ArrayAttr>(kVectorLayoutFetcherStorageAttrName);
+  assert(signatureAttr && "Op should have a signature attribute.");
+  assert(signatureAttr.size() == 2 && "Malformed signature attribute.");
+
+  ArrayAttr operandsAttr = dyn_cast<ArrayAttr>(signatureAttr[0]);
```

Comment:
Could we look up operand signature by looking up the attribute of its defining op?

Reply:
I personally prefer keeping a distribution signature for each operand. It makes more sense to me with patterns, because the signature makes the pattern self contained.